Use different ports for the docker containers in the matrix tests

IanCal opened the pull request:
This change lets the docker containers pick a free port and should not therefore clash with the development server you may be running at the same time.

One major benefit of this is that you can now run the tests in parallel, I've tested up to 32 in parallel.

This has highlighted several more tests that don't always pass, at least one with room objectives relates to some issue with the application rather than the test.

Containers can be started with a specific name and port mapping (so that the change is transparent to the rest of the system). If not then each container will get a unique name (matching the temp directory name that is created for them) and docker will assign a free port. The port mapping is taken from docker and passed through into the tests.

There are two key changes elsewhere:

1. The application route now, in development and test only, allows for changing the matrix URL using a querystring
2. The tests now have a helper that overrides goto and reload, ensuring the matrix URL remains set on the application when you navigate around or reload the page.

IanCal committed:
Support for custom matrix urls, single one for tests

IanCal committed:
Override page goto to add parameters, use port from the docker script

IanCal committed:
Use fixtures to handle setting up synapse per test

IanCal committed:
Parallel test running

IanCal committed:
Run synapse on random ports for each test

IanCal committed:
Remove logging

IanCal committed:
remove commented out work

IanCal committed:
minor cleanup & linting

IanCal committed:
Merge branch 'main' into matrix-test-url

IanCal committed:
Merge branch 'main' into matrix-test-url

IanCal committed:
Start synapse with explicit container name and port

IanCal committed:
Remove the playwrighttestmode parameter and allow optional overriding of matrix URL in test/dev

IanCal committed:
Merge branch 'main' into matrix-test-url

habdelra commented on packages/host/app/routes/application.ts:
```diff
@@ -8,6 +9,19 @@ interface Model {
+    // Override the matrix URL for testing
+    if (ENV.environment === 'test' || ENV.environment === 'development') {
+      if (transition.to.queryParams.matrixURL) {
+        ENV.matrixURL = transition.to.queryParams.matrixURL;
```
generally we set app state like this in a service. it's a bit awkward to actually set state at runtime in the config environment module. the config environment state is normally set during build time.

habdelra replied:
perhaps we should use the matrix service that already exists to set the matrix URL. we can set the matrix URL in the service from here when we see a query param, otherwise the matrix service will fallback to the env config.

habdelra replied:
arg, i was just looking thru the matrix service and i forgot that it lazily loads the matrix deps. calling the matrix service from this point would eagerly load the matrix deps. so nevermind--you are doing this the right way.

habdelra reviewed (approved):
looks good!

IanCal committed:
Drop concurrency back to 1

Just until we ensure the flakiness issues are solved, otherwise this exacerbates the issue.

IanCal committed:
Merge branch 'main' of github.com:cardstack/boxel

IanCal committed:
Merge branch 'matrix-test-url' of github.com:cardstack/boxel into matrix-test-url

IanCal committed:
Merge branch 'main' into matrix-test-url

habdelra commented on packages/matrix/helpers/index.ts:
```diff
@@ -12,6 +14,46 @@ interface LoginOptions {
+    newUrl.search = params.toString();
+
+    // Call the original goto function with the new URL
+    return originalGoto(newUrl.toString(), options);
```
Missing await here. Please try to be careful around missing async, it leads to flaky tests and super hard to debug.

habdelra commented on packages/matrix/helpers/index.ts:
```diff
@@ -12,6 +14,46 @@ interface LoginOptions {
+
+  // Patch the reload function
+  page.reload = async (options) => {
+    return page.goto(page.url(), options);
```
Missing await
